File: python/food-chain/food_chain.py
def recite(start_verse, end_verse):
    food_chain = ['fly', 'spider', 'bird', 'cat', 'dog', 'goat', 'cow', 'horse']
    song = []
    song_line_two = [
      '',
      '',
      'How absurd to swallow a ',
      'Imagine that, to swallow a ',
      'What a hog, to swallow a ',
      'Just opened her throat and swallowed a ',
      'I don\'t know how she swallowed a '
    ]

    for i in range(start_verse, end_verse+1):
        foods = food_chain[:i]
        foods.reverse()
        for food_idx, food in enumerate(foods):
            if food_idx == 0:
                song.append(f"I know an old lady who swallowed a {food}.")
                if i > 2 and i < len(food_chain):
                    song.append(f"{song_line_two[i-1]}{food}!")
            elif i > 2 and food_idx < i - 1 and i < len(food_chain):
                a_song = f"She swallowed the {foods[food_idx - 1]} to catch the {food}"
                if food_idx < i - 2:
                    a_song += '.'
                else:
                    a_song += ' that wriggled and jiggled and tickled inside her.'
                song.append(a_song)
            elif i < len(food_chain):
                if i == 2:
                    song.append("It wriggled and jiggled and tickled inside her.")
            
        
        if 1 < i < len(food_chain):
            song.append("She swallowed the spider to catch the fly.")
        if i == len(food_chain):
            song.append("She's dead, of course!")
        else:
            song.append("I don't know why she swallowed the fly. Perhaps she'll die.")
        if i < end_verse:
            song.append("")
        
    
    return song

File: ruby/food-chain/food_chain.rb
module FoodChain
  def self.song
    food_chain = ['fly', 'spider', 'bird', 'cat', 'dog', 'goat', 'cow', 'horse']
    eaten_food_chain = []
    song = ''
    song_line_2 = [
      '',
      '',
      'How absurd to swallow a ',
      'Imagine that, to swallow a ',
      'What a hog, to swallow a ',
      'Just opened her throat and swallowed a ',
      'I don\'t know how she swallowed a '
    ]


    food_chain.each_with_index do |f, i|
      song << "I know an old lady who swallowed a #{f}.\n"
      song << "#{song_line_2[i]}#{f}!\n" if i > 1 && i + 1 < food_chain.size
      eaten_food_chain << f
      if i + 1 < food_chain.size
        song << remaining_song(eaten_food_chain, i)
      else
        song << "She's dead, of course!\n"
      end
    end

    song
  end

  private

  def self.remaining_song(eaten_food_chain, index)
    partial_song = []
    previouse_eaten_food = nil

    eaten_food_chain.each_with_index do |f, i|

      partial_song.unshift("She swallowed the #{f} to catch the #{previouse_eaten_food}#{i==2 ? '' : ".\n"}") if i > 1

      partial_song.unshift("#{eaten_food_chain.size > 2 ? ' that' : "It"} wriggled and jiggled and tickled inside her.\nShe swallowed the #{f} to catch the #{previouse_eaten_food}.\n") if i == 1

      partial_song << "I don't know why she swallowed the #{f}. Perhaps she'll die.\n" if i == 0

      previouse_eaten_food = f
    end

    partial_song << "\n"
    partial_song.join
  end
end

module BookKeeping
  VERSION = 2
end